components/x-teaser-timeline/src/lib/transform.js

```diff
@@ -1,5 +1,51 @@
 import { getLocalisedISODate, getTitleForItemGroup, getDateOnly } from './date'
 
+/**
+ * @typedef { Array<string> | Array<Object> | Object | string } CustomSlotContent
+ * @typedef { Array<number> | number } CustomSlotPosition
+ */
+
+/**
+ * A news item (an article).
+ * @typedef {Object} Item
+ * @property {string} id e.g., '01f0b004-36b9-11ea-a6d3-9a26f8c3cba4'
+ * @property {string} title e.g.,'Europeans step up pressure on Iran over nuclear deal'
+ * @property {string} publishedDate e.g., '2020-01-14T11:10:26.000Z'
+ */
+
+/**
+ * Extra props added to an item in groups.
+ * @typedef {Object} ItemInGroupInfo
+ * @property {number} articleIndex
+ * @property {string} localisedLastUpdated e.g., '2020-01-14T11:10:26.000+00:00'
+ */
+
+/** A news item (an article) in a group of items.
+ * @typedef {(Item & ItemInGroupInfo)} ItemInGroup
+ */
+
+/**
+ * A list of news published on the same date.
+ * @typedef {Object} GroupOfItems
+ * @property {string} title e.g., 'Earlier Today'
+ * @property {string} date e.g., '2020-01-14'
+ * @property {Array<ItemInGroup>} items An array of news articles.
+ */
+
+/**
+ * @typedef {Object} PositionInGroup
+ * @property {number} group
+ * @property {number} index
+ */
+
```

Comment:
Praise: I found this really helpful... thank you!

Personally I tend to write  types in a separate `.d.ts` file (mainly because the syntax is nicer!) but it's actually kinda nice to have them right here too 😀